android/KMAPro/kMAPro/src/main/AndroidManifest.xml

```diff
@@ -191,6 +191,25 @@
           android:sspPrefix="download" />
 
       </intent-filter>
+
+      <intent-filter>
```

Comment:
Is this filter here in order to make sure that the download is handled by the MainActivity? Is it internal to the app?

Reply:
Correct. Previous intent-filters (above) look for URL patterns ending in `*.kmp` or ones using `keyman://download`.

> Is it internal to the app?

Can you clarify what you're asking about "internal to the app"?

Reply:
That is, if I visit that link from Chrome, will it open Keyman? (It shouldn't -- only the /keyboards/install/ links should).

Reply:
I've updated the intent filter  to only handle `/keyboards/install/`

atm, the staging keyman site isn't delivering the kmp file. Can I include `/go/package/download` in the meantime? (The screenshots I put in the PR description were from `/go/package/download`.)

Reply:
> atm, the staging keyman site isn't delivering the kmp file. Can I include `/go/package/download` in the meantime? (The screenshots I put in the PR description were from `/go/package/download`.)

I'm confused by this. Where are you expecting the keyman site to be delivering the kmp file from? Can you give the exact URLs you are loading and your expectations?

Reply:
From the [Install Process for Packages](https://docs.google.com/document/d/1rhgMeJlCdXCi6ohPb_CuyZd0PZMoSzMqGpv1A8cMFHY/edit#heading=h.mio0p3wdzdye) doc

### Keyboard Package Downloads
> Apps should download all packages from cloud using the following link pattern:
>
>https://keyman.com/go/package/download/package_id?[version=version][&platform=platform][&tier=tier][&bcp47=bcp47][&update=0|1]

I was able to get a kmp file
http://staging-keyman-com.azurewebsites.net/go/package/download/khmer_angkor?platform=android&tier=alpha&bcp47=km

Reply:
So I don't understand this bit:

> atm, the staging keyman site isn't delivering the kmp file.

Reply:
When the embedded browser goes to
https://staging-keyman-com.azurewebsites.net/keyboards/install/khmer_angkor it gets this webpage
![image](https://user-images.githubusercontent.com/7358010/87741672-4a030d00-c80f-11ea-83a2-752482f54908.png)

Meanwhile, the Keyman app is attempting to handle the file content (html) as if it were the served kmp file...

Reply:
I just need to clarify here: is this the embedded browser that you access through the Browser button in Keyman, and then you manually navigated to https://keyman.com/keyboards there? Or is this launched as a part of the Install Keyboard activity?

Reply:
Can we have a discussion on this? I don't think I am communicating the intended design clearly and I think that's part of why we are going around in circles a bit.

Reply:
Recap of our discussion: Using the Keyman app's KMPBrowserActivity for keyboard searches:
1. When the app displays the keyboard page (e.g. https://staging-keyman-com.azurewebsites.net/keyboards/khmer_angkor which is the 2nd screenshot above)
2. A click on the green install button is the URL 
https://staging-keyman-com.azurewebsites.net/keyboard/install/khmer_angkor 
(which redirects to 
https://staging-keyman-com.azurewebsites.net/keyboards/install/khmer_angkor )
3. But instead of having the activity display the keyboards/install link, the Keyman app parses the link and redirects to https://staging-keyman-com.azurewebsites.net/go/package/download/khmer_angkor?platform=android&tier=alpha
4. The intent filter triggers on /go/package and the MainActivity handles the kmp download/install...

Reply:
> 2\. A click on the green install button is the URL
>     https://staging-keyman-com.azurewebsites.net/keyboard/install/khmer_angkor
>     (which redirects to
>     https://staging-keyman-com.azurewebsites.net/keyboards/install/khmer_angkor )

Note that `/keyboard/install` link is a typo which will be removed soon (it will go directly to `/keyboards/install` -- the redirect is really for user typos not for this kinda of stoopidity!). So match only on `/keyboards/install/`.

> 3. But instead of having the activity display the keyboards/install link, the Keyman app parses the link and redirects to https://staging-keyman-com.azurewebsites.net/go/package/download/khmer_angkor?platform=android&tier=alpha
> 4. The intent filter triggers on /go/package and the MainActivity handles the kmp download/install...

Is there a need to do the redirect within the browser to `/go/package/...`? Can you not tell MainActivity to handle this directly without the hoop jumping and the intent filter?

Reply:
> ... the Keyman app parse the link and redirects to ...

Maybe redirect was the wrong word. The app generates the downloadURL to /go/package/download... and creates an  intent Action
```
 Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(downloadURL));
```

The intent filter triggers MainActivity to handle it (which then does the download/install)

Reply:
Gotcha. It'd be cleaner if the activity just a request in to some "`MainActivity.downloadPackage(packageId, bcp47)`" function which then does the work of constructing the URL and going off and downloading/installing, rather than generating the URL there. Is that possible?

Reply:
I'll put this PR back to wip to see if that's possible

Reply:
hm, I gave it a try. But I believe in order to go from one Activity to another, I need to create an Intent and start the other activity with it...

Reply:
OK, yes, you need to use an Intent, but I don't think you should be passing a URL in the Intent extra data, but rather parameterised variables of `packageId` and `bcp47`:
```
Intent intent = new Intent(this, MainActivity.class);
...
intent.putExtra(EXTRA_DOWNLOAD_PACKAGE_ID, packageId); 
intent.putExtra(EXTRA_DOWNLOAD_BCP47, bcp47);
```
My reasoning here is that the only place we should be constructing the URL is in the closest place where we are going to use it. That encapsulates knowledge of the URL structure in one place.

Reply:
I'll still need to pass a URL (Uri) for the intent filter to trigger the main app.

But I'll adjust KMPLink.getDownloadLink() to do the URL transform and return a URI with query parameters.